### PR TITLE
Fix race condition when setting MaxRetries on provisioning controller.

### LINKED ISSUE #2470: Race condition when setting MaxRetries on provisioning controller
**What happened**:
A race condition occurs when setting MaxRetries on provisioning controller.

**What you expected to happen**:
No race condition

**How to reproduce it (as minimally and precisely as possible)**:
GINKGO_ARGS=--race make test-integration

**Anything else we need to know?**:
```
WARNING: DATA RACE
  Write at 0x000003df1458 by goroutine 2233:
    sigs.k8s.io/kueue/test/integration/controller/admissionchecks/provisioning.init.func1.4.2()
        /home/prow/go/src/sigs.k8s.io/kueue/test/integration/controller/admissionchecks/provisioning/provisioning_test.go:1003 +0x70
    github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func3()
        /home/prow/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.19.0/internal/node.go:472 +0x2e
    github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
        /home/prow/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.19.0/internal/suite.go:894 +0x12b
  Previous read at 0x000003df1458 by goroutine 535:
    sigs.k8s.io/kueue/pkg/controller/admissionchecks/provisioning.(*Controller).syncOwnedProvisionRequest()
        /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/provisioning/controller.go:218 +0x7f6
    sigs.k8s.io/kueue/pkg/controller/admissionchecks/provisioning.(*Controller).Reconcile()
        /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/provisioning/controller.go:143 +0x8ed
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile()
        /home/prow/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:119 +0x1a1
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler()
        /home/prow/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:316 +0x59a
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem()
        /home/prow/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:266 +0x338
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /home/prow/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:227 +0xb2
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/2468/pull-kueue-test-integration-main/1804230951551635456

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
To fix race condition when setting MaxRetries on provisioning controller.

#### Which issue(s) this PR fixes:

Fixes #2470

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Review by mszadkow — APPROVED** [review]
lgtm

File: pkg/controller/admissionchecks/provisioning/controller.go [hunk 1]
```diff
@@ -60,25 +63,54 @@ var (
...
+			controller.waitForProvisioningRequestReady.maxRetries = *cfg.MaxRetries
+		}
+		if cfg.MinBackoffSeconds > 0 {
+			controller.waitForProvisioningRequestReady.minBackoffSeconds = cfg.MinBackoffSeconds
+		}
+		if cfg.MaxBackoffSeconds > 0 {
+			controller.waitForProvisioningRequestReady.maxBackoffSeconds = cfg.MaxBackoffSeconds
+		}
+	}
+}
```
Comment (trasc) [inline pkg/controller/admissionchecks/provisioning/controller.go:112]:
```suggestion
func WithWaitForProvisioningMaxRetries(v int32) Option {
	return func(controller *Controller) {
		controller.waitForProvisioningRequestReady.maxRetries =  &v 
	}
}

func WithWaitForProvisioningMinBackoffSeconds(v int32) Option {
	return func(controller *Controller) {
		controller.waitForProvisioningRequestReady.minBackoffSeconds = v
	}
}

func WithWaitForProvisioningMaxBackoffSeconds(v int32) Option {
	return func(controller *Controller) {
		controller.waitForProvisioningRequestReady.maxBackoffSeconds = v
	}
}
```

**Review by trasc — COMMENTED** [review]
just one nit, otherwise 

/lgtm

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 1 above)
Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/provisioning/controller.go]: Done

File: pkg/controller/admissionchecks/provisioning/controller.go [hunk 2]
```diff
@@ -60,25 +63,41 @@ var (
...
 var _ reconcile.Reconciler = (*Controller)(nil)
 
+// Option configures the provisioning controller.
+type Option func(*Controller)
+
+func WithWaitForProvisioningMaxRetries(maxRetries int32) Option {
+	return func(controller *Controller) {
+		controller.waitForProvisioningRequestReady.maxRetries = maxRetries
+	}
+}
```
Comment (tenzen-y) [inline pkg/controller/admissionchecks/provisioning/controller.go:94]:
```suggestion
type Option func(*options)

type options struct {
	waitForProvisioningRequestReady *waitForProvisioningRequestReady
}

func WithWaitForProvisioningMaxRetries(maxRetries int32) Option {
	return func(o *options) {
		o.waitForProvisioningRequestReady = maxRetries
	}
}
```

Could you avoid exposing unnecessary fields in the Controller?
Additionally, we can avoid null pointers by such dedicated typed `options`.

**Comment by tenzen-y** [comment]
@mbobrovskyi could you rebase this PR?

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 2 above)
Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/provisioning/controller.go]: Done

**Comment by mbobrovskyi** [comment]
@alculquicondor @tenzen-y 

> There are no nodes that your pod can schedule to - check your requests, tolerations, and node selectors (0/46 nodes are available: 3 node(s) had untolerated taint {node-group: stable}, 43 Insufficient cpu. preemption: 0/46 nodes are available: 3 Preemption is not helpful for scheduling, 43 No preemption victims found for incoming pod..)

Is it possible some how to increase limitation?

**Comment by alculquicondor** [comment]
> Is it possible some how to increase limitation?

No, that's the k8s-infra cluster. The only thing we can do is reduce the size of our test pod, but we usually need all those resources to avoid timeouts.

File: pkg/controller/admissionchecks/provisioning/controller.go [hunk 3]
```diff
@@ -60,25 +63,45 @@ var (
...
-	MinBackoffSeconds int32 = 60
-)
-
 type provisioningConfigHelper = admissioncheck.ConfigHelper[*kueue.ProvisioningRequestConfig, kueue.ProvisioningRequestConfig]
 
 func newProvisioningConfigHelper(c client.Client) (*provisioningConfigHelper, error) {
 	return admissioncheck.NewConfigHelper[*kueue.ProvisioningRequestConfig](c)
 }
 
+type waitForProvisioningRequestReadyConfig struct {
```
Comment (alculquicondor) [inline pkg/controller/admissionchecks/provisioning/controller.go:72]: A bit too much indirection. Just put these values inside the `options` struct.

File: pkg/controller/admissionchecks/provisioning/controller.go [hunk 4]
```diff
@@ -60,25 +63,45 @@ var (
...
 var _ reconcile.Reconciler = (*Controller)(nil)
 
+type options struct {
+	waitForProvisioningRequestReady *waitForProvisioningRequestReadyConfig
+}
+
+// Option configures the provisioning controller.
+type Option func(*options)
+
+func WithWaitForProvisioningMaxRetries(maxRetries int32) Option {
```
Comment (alculquicondor) [inline pkg/controller/admissionchecks/provisioning/controller.go:94]:
```suggestion
func WithMaxRetries(maxRetries int32) Option {
```

same below

File: pkg/controller/admissionchecks/provisioning/controller.go [hunk 5]
```diff
@@ -50,7 +50,10 @@ import (
 )
 
 const (
-	objNameHashLength = 5
+	DefaultMaxRetries        int32 = 3
+	DefaultMinBackoffSeconds int32 = 60   // 1 min
+	DefaultMaxBackoffSeconds int32 = 1800 // 30 min
```
Comment (alculquicondor) [inline pkg/controller/admissionchecks/provisioning/controller.go:55]: I think this is possible now

```suggestion
	defaultMaxRetries        int32 = 3
	defaultMinBackoffSeconds int32 = 60   // 1 min
	defaultMaxBackoffSeconds int32 = 1800 // 30 min
```

File: test/integration/controller/admissionchecks/provisioning/suite_test.go [hunk 6]
```diff
@@ -53,33 +57,39 @@ func TestProvisioning(t *testing.T) {
 	)
 }
 
-func managerSetup(mgr manager.Manager, ctx context.Context) {
-	err := indexer.Setup(ctx, mgr.GetFieldIndexer())
-	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+func managerSetup(opts ...provisioning.Option) framework.ManagerSetup {
+	return func(mgr manager.Manager, ctx context.Context) {
```
Comment (alculquicondor) [inline test/integration/controller/admissionchecks/provisioning/suite_test.go:61]: Maybe just this?
```suggestion
func managerSetup(mgr manager.Manager, ctx context.Context, opts ...provisioning.Option) {
```

Comment (mbobrovskyi) [reply test/integration/controller/admissionchecks/provisioning/suite_test.go:61]: Unfortunately no. The `ManagerSetup` that using on `fwk.RunManager` required only this two fields.

```go
type ManagerSetup func(manager.Manager, context.Context)
```

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 3 above)
Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/provisioning/controller.go]: Thanks. Done.

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 4 above)
Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/provisioning/controller.go]: Done

File: pkg/controller/admissionchecks/provisioning/controller.go (hunk 5 above)
Comment (mbobrovskyi) [reply pkg/controller/admissionchecks/provisioning/controller.go]: Done

File: test/integration/controller/admissionchecks/provisioning/provisioning_test.go [hunk 7]
```diff
@@ -47,24 +46,25 @@ const (
...
+			provisioning.WithWaitForProvisioningMaxRetries(maxRetries),
+			provisioning.WithWaitForProvisioningMinBackoffSeconds(minBackoffSeconds),
+		))
 	})
 
-	ginkgo.AfterAll(func() {
+	ginkgo.AfterEach(func() {
 		fwk.Teardown()
+		maxRetries = provisioning.DefaultMaxRetries
+		minBackoffSeconds = provisioning.DefaultMinBackoffSeconds
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply test/integration/controller/admissionchecks/provisioning/provisioning_test.go]: You right. Thanks. Done.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**